src/launch/mod.rs

```diff
@@ -192,3 +194,224 @@ impl codicon::Encoder<()> for Measurement {
         writer.save(self)
     }
 }
+
+bitflags! {
+    /// Configurable SNP Policy options.
+    #[derive(Default, Deserialize, Serialize)]
+    pub struct SnpPolicyFlags: u16 {
+        /// Enable if SMT is enabled in the host machine.
+        const SMT = 1;
+
+        /// If enabled, association with a migration agent is allowed.
+        const MIGRATE_MA = 1 << 2;
+
+        /// If enabled, debugging is allowed.
+        const DEBUG = 1 << 3;
+    }
+}
+
+/// Describes a policy that the AMD Secure Processor will
+/// enforce.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, Deserialize, Serialize)]
+pub struct SnpPolicy {
+    /// The various policy optons are encoded as bit flags.
+    pub flags: SnpPolicyFlags,
+
+    /// The desired minimum platform firmware version.
+    pub minfw: Version,
+}
+
+impl SnpPolicy {
+    /// Convert a Policy to it's u64 counterpart.
+    pub fn as_u64(&self) -> u64 {
+        let mut val: u64 = 0;
+
+        let minor_version = u64::from(self.minfw.minor);
+        let mut major_version = u64::from(self.minfw.major);
+
+        /*
+         * According to the SNP firmware spec, bit 1 of the policy flags is reserved and must
+         * always be set to 1. Rather than passing this responsibility off to callers, set this bit
+         * every time an ioctl is issued to the kernel.
+         */
+        let flags = self.flags.bits | 0b10;
+        let mut flags_64 = u64::from(flags);
+
+        major_version <<= 8;
+        flags_64 <<= 16;
+
+        val |= minor_version;
+        val |= major_version;
+        val |= flags_64;
+        val &= 0x00FFFFFF;
+
+        val
+    }
+}
+
+/// Encapsulates the various data needed to begin the launch process.
+#[derive(Clone, Debug, Default, PartialEq, Eq, Deserialize, Serialize)]
+pub struct SnpStart<'a> {
+    /// The userspace address of the region to be encrypted.
+    pub uaddr: Option<&'a [u8]>,
+
+    /// Describes a policy that the AMD Secure Processor will enforce.
+    pub policy: SnpPolicy,
+
+    /// Indicates if this guest is associated with a migration agent. Otherwise 0.
+    pub ma_en: bool,
+
+    /// Indicates that this launch flow is launching an IMI for the purpose of guest-assisted migration.
+    pub imi_en: bool,
+
+    /// Hypervisor provided value to indicate guest OS visible workarounds.The format is hypervisor defined.
+    pub gosvw: [u8; 16],
+}
+
+impl<'a> SnpStart<'a> {
+    /// Encapsulate all data needed for the SNP_LAUNCH_START ioctl.
+    pub fn new(uaddr: Option<&'a [u8]>, policy: SnpPolicy, imi_en: bool, gosvw: [u8; 16]) -> Self {
+        Self {
+            uaddr,
+            policy,
+            ma_en: uaddr.is_some(),
```

Comment:
This is incorrect... not every guest is a migration agent!

Reply:
Ah, sorry, this is the migration assistent uaddr.

Reply:
That wasn't clear

Reply:
should probably renamed to `ma_uaddr`

Reply:
I see how that can be confusing and agree. Changed it to `ma_uaddr`